### DISCUSSION TITLE
Chinese characters with Text2D

### DISCUSSION THREAD

**Discussion by ZiguoAtGitHub** [marcomusy/vedo#715] [Q&A]
Hi, All
I copied the chinese font file into vedo font directory and coded the Text2D(...,font='font_file.ttf',...). But it didn't work.
Adding the path to the font_file as "c:\windows\font\font_file.ttf" did not work neither.

Call for help~
Thank you All!

**Comment by marcomusy** [comment]
Hi @ZiguoAtGitHub 
that is strange - it works fine on my ubuntu system..
Have you tried `examples/pyplot/caption.py`, it supports Chinese chars:
`font="LogoType"`

**Comment by ZiguoAtGitHub** [comment]
### Hi @marcomusy ,
### First of all, Thanks for your reply.
### Yes. it is very strange. I just download the caption.py that you refered and run  as "python caption.py " in cmd(my pc env is win10, python 3.9.10), here is error.
![image](https://user-images.githubusercontent.com/116484589/197393588-51b25d27-c64c-4c7b-ba18-f6d2c634d351.png)
### The problem looks like caused by missing LogoType.ttf, doesn't it?
### Then, I replace the font_file with STSONG.ttf which is found in system font directory and the error is:
![image](https://user-images.githubusercontent.com/116484589/197394273-2634489d-f069-41e4-9885-6cc3fde7ecef.png)
### And the vedo plot window is 
![image](https://user-images.githubusercontent.com/116484589/197394289-b713800e-b036-495c-8310-7e32cc567d49.png)

### Pls give me some help, Thank you!
### BTW: Vedo is amazing to me, Great job!

**Reply by marcomusy** [reply]
Hi @ZiguoAtGitHub 
I'm a bit confused by the error message.. To be on the same page, can you try
`pip install -U git+https://github.com/marcomusy/vedo.git`

then run again `caption.py` example.
Make sure you have an internet connection as the `LogoType` font is downloaded on demand.

> BTW: Vedo is amazing to me, Great job!

Thnaks!

**Reply by ZiguoAtGitHub** [reply]
### Hi @marcomusy ,
### It is successful as I run `captio.py` example on another pc (windows10 and python 3.8.10-amd64) this morning.
![image](https://user-images.githubusercontent.com/116484589/197429530-294e8998-69ad-406d-b26a-3025ca091f26.png)
### But if I change the `font="LogoType"`(LogoType is a traditional chinese font) to `font="STFANGSO.TTF"`(STFANGSO.ttf is a simplified chinese font and copied to the vedo font directory), the problem is still there.
![image](https://user-images.githubusercontent.com/116484589/197431010-82898e3c-6332-4268-88d9-fd3f03e5b35e.png)
![image](https://user-images.githubusercontent.com/116484589/197431011-3ef31c9f-c9a1-4565-a487-70568676ed24.png)

### The followings are codes I modified:
```python
#encoding:utf-8
"""Attach a 2D caption to an object"""
from vedo import Cone, Axes, show

cone = Cone().rotate_x(30).rotate_z(20).c('steelblue')

txt  = "Japanese\nこれは青い円錐形です\n"
txt += "Chinese\n这是一个蓝锥\n"
txt += "Russian\nЭто синий конус\n"
txt += "English\nThis is a blue cone"

cone.caption(txt, size=(0.4,0.3), font="STFANGSO.ttf", c='lb')

axes = Axes(
    cone,
    xtitle='マイクロメートル単位のx軸',
    ytitle='y軸にも長い説明があります',
    ztitle=u'Z轴',
    title_font='STFANGSO.ttf',
    text_scale=1.5,
    c='white',
)

show(cone, axes, __doc__, viewup='z', bg='k', bg2='bb').close()
```

### Many Thanks !

**Reply by marcomusy** [reply]
The error message is actually normal because the chars that show on the axes are 3D so they need an extra step to work out a polygonal representation of them!
I found online a file with the same name ..but it contains 40k different symbols so I would need to know from you which is the range of IDs that you wish to have (e.g. "I want symbols from 3000 to 6000").
I attach a file with symbols and corresponding IDs.

Also I'm not sure if the file I found online is free and if it is the correct one.

[font_chinese.txt.gz](https://github.com/marcomusy/vedo/files/9852242/font_chinese.txt.gz)

**Reply by ZiguoAtGitHub** [reply]
Hi @marcomusy ,
Thank you for the reply and I think I get the answer of my problem. Actually, all characters are plotted as geometry shapes in vedo, including Text2D and other similar functionalities. Am I right?
And I suppose the extra works would be huge to generate 3D polygons for representing the symbols of a new font. So I need some time to collect the symbols' ID which I do need and control the numbers of IDs as few as possible. The ID of the symbol in font_chinese.txt looks like the symbol's unicode number, isn't it?

Thanks for your kindly help!

**Reply by marcomusy** [reply]
> all characters are plotted as geometry shapes in vedo, including Text2D and other similar functionalities. Am I right?

well in the case of Text2D we take the shortcut and use the tff file directly..

> the extra works would be huge to generate 3D polygons

it takes 2 minutes, but the output file for STFANGSO is 190 MB :) so we need to reduce the nr of desired symbols

> The ID of the symbol in `font_chinese.txt` looks like the symbol's unicode number, isn't it?

Yes, t's the number right after the symbol, eg:
```
Python 3.9.13 (main, Aug 25 2022, 23:26:10) 
[GCC 11.2.0] :: Anaconda, Inc. on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> chr(33)
'!'
```

**Reply by ZiguoAtGitHub** [reply]
Hi @marcomusy ,
As you mentioned `well in the case of Text2D we take the shortcut and use the tff file directly..,` , I took a try about plotting chinese chars by Text2D (the codes are `kk=Text2D("This is a chinese words:简体中文(simplified chinese)", font='c:\\windows\\fonts\\STFANGSO.TTF', pos='top-left', c='red3')`. Here is the result.
![image](https://user-images.githubusercontent.com/116484589/197679159-167a31c1-c98e-403f-b3a4-1876767704e7.png)
It looks like the parameter font='c:\\windows\\fonts\\STFANGSO.TTF' only take effects on english characters and the chinese characters are skipped. Did I miss something?

Wating for your reply. 
Thanks a lot!!

**Reply by marcomusy** [reply]
Hi @ZiguoAtGitHub I was hoping that was a problem with my version of STFANGSO but infact you got the same...
I think it must be a problem with the upstream VTK, I will try to send a request to the developers to see if they understand what happens with STFANGSO.ttf (it still works in 3D though in my local tests).
I'd love to see chinese chars in `vedo`!
Thanks for your patience

**Reply by ZiguoAtGitHub** [reply]
Hi @marcomusy ,
Thanks for your reply!
`to see chinese chars in vedo!`
I am looking forward to that also. :)

Thank you again for your great job, it does give me more imaginary space for my job.

**Reply by marcomusy** [reply]
I've been playing around with online fonts (I love fonts) and this is what i've come up with:
```bash
pip install -U git+https://github.com/marcomusy/vedo.git
```

Then
```python
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
from vedo import Axes, Text2D, Text3D, show, settings

settings.default_font = "Housekeeper" # ~28MB
# settings.default_font = "Wananti"    # ~74MB

txt = ( "简体字是在中国大陆、"
        "马来西亚和新加坡使用的标准化汉字,\n"
        "由通用标准汉字表规定。\n"
        "与繁体字一起, \n"
        "它们是当代汉语书面语言的两个标准字符集之一。")


t2d = Text2D(txt, s=2)
t3d = Text3D(txt, c='blue2').bc("red5")

axes = Axes(
    t3d,
    xtitle='x 简体中文',
    ytitle='y 轴',
    text_scale=1.5,
)

show(t2d, t3d, axes, size=(1500,500)).close()
# for t in txt: print(t, '=', ord(t))
```
```
Housekeeper
https://chinesefontdesign.com/font-housekeeper-flattening-bold-figure-chinese-font-simplified-chinese-fonts.html
```

![Screenshot from 2022-10-25 18-02-53](https://user-images.githubusercontent.com/32848391/197828192-33227e64-fe40-4b6d-ac8e-44388a465729.png)

```
Wananti
https://chinesefontdesign.com/good-night-retro-chinese-font-simplified-chinese.html
```
![Screenshot from 2022-10-25 18-03-27](https://user-images.githubusercontent.com/32848391/197828188-89385ad8-9e3c-4c24-bd89-95a59482b6bd.png)

they are pretty large files and the final result is not 100% perfect.... 
How does it look?

**Reply by ZiguoAtGitHub** [reply]
Hi @marcomusy ,
I replicated your results. IT IS WONDERFUL!
![image](https://user-images.githubusercontent.com/116484589/197911806-0678966a-a7c7-44e0-89ab-15e83cfa7cd4.png)
Thanks for your grateful work.
